### PR TITLE
Issue 865: Add 404 routes to nginx config file

### LINKED ISSUE #865: Add 404 routes to nginx config file
**Description:**

We recently ran into an issue in the OWASP Scan

to fix this we would be adding the same routing rules we have using react-router to nginx. 
The main thing to be aware of will be that any new routes after we do this will have to be added 
in two places (tdrs-frontend/src/components/Routes.js and tdrs-frontend/nginx/nginx.conf)

we have to explicitly declare the valid paths, and treat everything else as 404, then implement a 
404 page, either a static html file, or as a react component, I would be leaning towards the react component.

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [x] High risk zap scan item is no longer present
- [x] Routes we have not explicitly provided react routes for, should serve a 404
- [x] When the react app detects the page has been served with a 404 status, it should display a Not Found page instead of a blank one
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [x] Update nginx.conf file to be more selective about what pages are valid
- [x] Update nginx.conf file to emit a 404 response
- [x] Add 404 react component
- [x] Run Testing Checklist and confirm all tests pass
- [x] @ADPennington to update POAM

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

+ In the future, if we add routes, we will have to add them to both the routes.jsx file and the nginx.conf file

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

+ [No match react route example](https://reactrouter.com/web/example/no-match)

### PR DESCRIPTION
**Pull request by riatzukiza**
### Summary of Changes

Resolves #865 
owasp saw that any endpoint was valid, and it gave us a high priority warning for it. Technically the issue that it saw wasn't there, but it believed that it was because a url that would usually represent a vunerability was returning 200.

We had no concept of a 404 in our code, this PR adds a 404 location to nginx, and a very basic react 404 page.

#### Todos not in scope
* Design a 404 page that better fits our app: https://github.com/raft-tech/TANF-app/issues/927
* Implement above 404 page design: https://github.com/raft-tech/TANF-app/issues/928

### How to Test

1. Deploy this PR
2. Go to any url that we haven't explicitly defined to see a 404 page
3. test existing features to see that they do not produce 404 messages

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  - [x] High risk zap scan item is no longer present
  - [x] Routes we have not explicitly provided react routes for, should serve a 404
  - [x] When the react app detects the page has been served with a 404 status, it should display a Not Found page instead of a blank one
  - [x] Testing Checklist has been run and all tests pass
  - [x] README is updated, if necessary

As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [x] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?
+ [x] If this PR introduces frontend code, is that code documented both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/design-issue-template.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ This is a follow-on to #
 AC: 
 - [ ] The problem we are trying to solve is clearly stated, as are questions we still have 
 - [ ] The design is consistent with the team’s past decisions, or a change is clearly documented 
+- [ ] The design is reviewed alongside [our design principles](https://github.com/HHS/TANF-app/wiki/Design-principles) and any friction is documented and discussed with the team
```
Comment (jorgegonzalez) [inline .github/ISSUE_TEMPLATE/design-issue-template.md:17]:
```suggestion
```

File: .github/ISSUE_TEMPLATE/dev-issue-template.md [hunk 2]
```diff
@@ -17,6 +17,7 @@ _Create a list of functional outcomes that must be achieved to complete this iss
 - [ ] _Outcome 1_
 - [ ] _Outcome 2_
 - [ ] _Outcome 3_
+- [ ] Work is reviewed alongside [our design principles](https://github.com/HHS/TANF-app/wiki/Design-principles) and any friction is documented and discussed with the team
```
Comment (jorgegonzalez) [inline .github/ISSUE_TEMPLATE/dev-issue-template.md:20]:
```suggestion
```

File: .github/ISSUE_TEMPLATE/research-issue-template-.md [hunk 3]
```diff
@@ -12,7 +12,7 @@ Description:
 
 AC:
 
-- [ ] [Draft research plan]/[user interviews]/[synthesis] is complete.
+- [ ] [Draft research plan]/[user interviews]/[synthesis] is complete.
```
Comment (jorgegonzalez) [inline .github/ISSUE_TEMPLATE/research-issue-template-.md:15]:
```suggestion
- [ ] [Draft research plan]/[user interviews]/[synthesis] is complete.
```

File: tdrs-frontend/nginx/nginx.conf [hunk 4]
```diff
@@ -1,15 +1,36 @@
 server {
-  listen 80;
+    listen 3005;
```
Comment (jorgegonzalez) [inline tdrs-frontend/nginx/nginx.conf:2]:
```suggestion
    listen 80;
```

**Comment by carltonsmith** [comment]
@riatzukiza This looks great, but it would be good to have a test that shows a 404 when applicable

**Review by carltonsmith — DISMISSED** [review]
Almost ready to move up once we have some tests. Thanks!

**Review by jtwillis92 — APPROVED** [review]
Pulled this down locally and confirmed the expected behavior. Additionally the high risk item is gone from the OWASP ZAP report as intended - remaining alerts are false positives and will be resolved with https://github.com/raft-tech/TANF-app/issues/879

**Comment by jtwillis92** [comment]
Since [the buildpacks PR](https://github.com/raft-tech/TANF-app/pull/881) was merged before this one the Dockerfile changes made in this PR are no longer being used. Taking this out of QASP for now while we consolidate the nginx configuration with the buildpack version.

File: tdrs-frontend/nginx/buildpack.nginx.conf [hunk 5]
```diff
@@ -20,8 +20,33 @@ http {
     server {
         listen {{port}};
 
-        root public;
-        index index.html index.htm;
-        try_files $uri $uri/ /index.html;
+        location = /edit-profile {
```
Comment (jtwillis92) [inline tdrs-frontend/nginx/buildpack.nginx.conf:23]: Would there be any issues with combining these two nginx configurations into one file?

Having to maintain these URLs in both of these files along with the React Router code seems like overkill.

**Review by jtwillis92 — CHANGES_REQUESTED** [review]
On the deployed site this is now resulting in the nginx 404 page for every path:
Screenshot from 2021-06-03 15-14-03

File: tdrs-frontend/nginx/buildpack.nginx.conf (hunk 5 above)
Comment (riatzukiza) [reply tdrs-frontend/nginx/buildpack.nginx.conf:23]: We could have this file include the other file inside of the server directive I think, the buildpacks have to declare a server, but the docker config seems to be included under a server block from inside the container.

**Review by jtwillis92 — APPROVED** [review]
LGTM, tested locally and on the deployed site with no issues.

**Comment by ADPennington** [comment]
👍🏾 test results:

- [x] checked defined frontend routes since this warning appeared in frontend owasp report:
✔️ https://tdp-frontend.app.cloud.gov/data-files
✔️ https://tdp-frontend.app.cloud.gov/edit-profile
✔️ https://tdp-backend.app.cloud.gov/admin/
❌ https://tdp-frontend.app.cloud.gov/welcome
(do we have a ticket to define and improve on this page?)
404_welcome

- [x] checked undefined frontend routes 
404bad

**Review by ADPennington — APPROVED** [review]
thanks @riatzukiza 👍🏾 👍🏾 

@abottoms-coder this is ready to merge! A couple of things to highlight from this review:
- no gov a11y review was conducted for this, so we're not too sure if the 404 pages are accessible but the follow-on issue @jtwillis92 created could include an accessibility check as a task.  
- welcome page route isnt defined yet, and this could also be addressed as part of follow-on work to build that page out
- i added a tick mark next to frontend owasp findings below that were not appearing prior to scans taking for ATO, so it will be helpful for us to get a better sense for what these mean and if existing tickets will address them if remediation is needed. 
---

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
 - [x]  High risk zap scan item is no longer present
_https://7588-281707402-gh.circle-artifacts.com/0/tdrs-frontend/reports/owasp_report.html_
 - [x]  Routes we have not explicitly provided react routes for, should serve a 404
 - [x]  When the react app detects the page has been served with a 404 status, it should display a Not Found page instead of a blank one
 - [x]  Testing Checklist has been run and all tests pass

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?


### Deliverable 4: Accessible

_not checked, but #928 created, which could include confirming that this design is accessible_ 

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces frontend code, is that code documented both inline and overall?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
_good to see no more high warnings! I am noticing some medium, low, and informational frontend findings that can be addressed in a follow-on since these were not introduced with this work and #879 and #909 may be addressing at least some of these_
frontend_owasp

**Pull request merged**